modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/entity/SearchHistoryItem.kt

```diff
@@ -0,0 +1,44 @@
+package au.com.shiftyjelly.pocketcasts.models.entity
+
+import androidx.room.ColumnInfo
+import androidx.room.Embedded
+import androidx.room.Entity
+import androidx.room.Index
+import androidx.room.PrimaryKey
+import java.util.Date
+
+@Entity(
+    tableName = "search_history",
+    indices = [
+        Index(value = ["term"], unique = true),
+        Index(value = ["podcast_uuid"], unique = true),
+        Index(value = ["folder_uuid"], unique = true),
+        Index(value = ["episode_uuid"], unique = true),
+    ]
+)
+data class SearchHistoryItem(
+    @PrimaryKey(autoGenerate = true) @ColumnInfo(name = "_id") var id: Long? = null,
+    @ColumnInfo(name = "modified") var modified: Long = System.currentTimeMillis(),
+    @ColumnInfo(name = "term") var term: String? = null,
+    @Embedded(prefix = "podcast_") var podcast: Podcast? = null,
+    @Embedded(prefix = "folder_") var folder: Folder? = null,
+    @Embedded(prefix = "episode_") var episode: Episode? = null,
+) {
+    data class Podcast(
+        val uuid: String,
+        val title: String,
+        val author: String,
+    )
+    data class Folder(
+        val uuid: String,
+        val title: String,
+        val color: Int,
+        val podcastIds: String,
+    )
+    data class Episode(
+        val uuid: String,
+        val title: String,
+        val publishedDate: Date,
+        val duration: Double
```

Comment:
These fields might change based on how an episode's recent search (logged when user taps an item from the episode search results) is displayed.

Reply:
Am I understanding correctly that the idea here is that each Search History Item would be for either a `Podcast`,`Folder`, or `Episode`, so the embedded fields for only one of those three would not be `null` in any given `SearchHistoryItem`?

If that's right, it might be worth considering at some point using a sealed class for this so we don't have so many nullable fields to deal with in our code. I'm not sure I'd do that at the db level necessarily (what you have here is nice and straightforward), but perhaps when we start using this part of the db, it would be worth creating a class for accessing the database that would convert all of the entries to/from an appropriate sealed class. Very rough pseudo code of what I'm thinking of:

<details>
<summary>pseudo code</summary>

``` kotlin
sealed class SearchHistoryEntry(
    val modified: Long,
    /* ... */
) {

    class Podcast(
        modified: Long,
        val author: String,
        // ...
        ) : SearchHistoryEntry(modified /* ... */)

    class Folder(
        modified: Long,
        val color: Int
        // ...
    ): SearchHistoryEntry(modified /* ... */)

    class Episode(
        modified: Long,
        val publishedDate: Date,
        // ...
    ): SearchHistoryEntry(modified, /* ... */)
}

class SearchHistoryDbUtil(private val searchHistoryDao: SearchHistoryDao) {
    suspend fun insert(entry: SearchHistoryEntry) {
        val item: SearchHistoryItem = when (entry) {
            is SearchHistoryEntry.Episode -> // create SearchHistoryItem from SearchHistoryEntry.Episode
            is SearchHistoryEntry.Folder -> // create SearchHistoryItem from SearchHistoryEntry.Folder
            is SearchHistoryEntry.Podcast -> // create SearchHistoryItem from SearchHistoryEntry.Podcast
        }
        searchHistoryDao.insert(item)
    }

    suspend fun findAll(limit: Int = 10): List<SearchHistoryEntry> {
        val entries = searchHistoryDao.findAll(limit)
        return entries.map { it.toEntry() }
    }

    private fun SearchHistoryItem.toEntry(): SearchHistoryEntry {
        // create appropriate SearchHistoryEntry from SearchHistoryItem
    }
}
```

</details>

We may figure out that this is actually a bad idea once we start implementing this, but I just wanted to mention it as a possibility in case you might find that it would actually be useful as you start using the db.

Reply:
> Am I understanding correctly that the idea here is that each Search History Item would be for either a Podcast,Folder, or Episode, so the embedded fields for only one of those three would not be null in any given SearchHistoryItem?

That's correct.

> If that's right, it might be worth considering at some point using a sealed class for this so we don't have so many nullable fields to deal with in our code. I'm not sure I'd do that at the db level necessarily (what you have here is nice and straightforward), but perhaps when we start using this part of the db, it would be worth creating a class for accessing the database that would convert all of the entries to/from an appropriate sealed class. Very rough pseudo code of what I'm thinking of:

I did try to add relations and different tables for each sub-entity and foreign key and all, but it looked complicated. So I didn't use it at the db level to not over-optimize it considering that embedded objects are small. I like your suggestion to use a sealed class at another point and will use it in another PR.

I might also add a query to keep a limited number of records in the table at a later point.